### PR TITLE
Update to JupyterLab 4

### PR DESCRIPTION
**Pull request by lahwaacz**
Output of `python -m jupyterlab.upgrade_extension .`:

```
/home/lahwaacz/stuff/jupyterlab-pytutor/.venv/lib/python3.11/site-packages/copier/template.py:146: OldTemplateWarning: This template was designed for Copier 7.1.0, but your version of Copier is 8.1.0. You could find some incompatibilities.
  warn(

Copying from template version 4.1.0
    create  .
    create  package.json
    create  jupyterlab-pytutor
    create  jupyterlab-pytutor/__init__.py
    create  .copier-answers.yml
    create  schema
    create  schema/plugin.json
    create  tsconfig.json
    create  style
    create  style/index.js
    create  style/base.css
    create  style/index.css
    create  src
    create  src/index.ts
    create  setup.py
    create  pyproject.toml
    create  install.json
    create  RELEASE.md
    create  README.md
    create  LICENSE
    create  CHANGELOG.md
    create  .yarnrc.yml
    create  .prettierignore
    create  .gitignore
    create  .github
    create  .github/workflows
    create  .github/workflows/check-release.yml
    create  .github/workflows/publish-release.yml
    create  .github/workflows/prep-release.yml
    create  .github/workflows/enforce-label.yml
    create  .github/workflows/build.yml

Overwrite scripts in package.json? [n]: y
overwrite ".prettierignore"? [y]: y
overwrite "README.md"? [n]:
overwrite "RELEASE.md"? [y]:
overwrite "tsconfig.json"? [y]:
overwrite ".gitignore"? [y]:
overwrite "pyproject.toml"? [y]:
overwrite "setup.py"? [y]:
overwrite "LICENSE"? [n]:
overwrite "install.json"? [n]:
overwrite "CHANGELOG.md"? [n]:
overwrite "src/index.ts"? [n]:
overwrite ".github/workflows/check-release.yml"? [y]: overwrite ".github/workflows/build.yml"? [y]:
overwrite "schema/plugin.json"? [n]:
** Updated scripts in package.json
** DELETED .eslintrc.js
** DELETED .eslintignore
** DELETED .prettierrc
** DELETED .stylelintrc
** skipped _temp_extension/README.md
** skipped _temp_extension/LICENSE
** skipped _temp_extension/install.json
** skipped _temp_extension/CHANGELOG.md
** skipped _temp_extension/src/index.ts
** skipped _temp_extension/schema/plugin.json
** Remove _temp_extensions directory when finished
```

### CODE REVIEW

**Comment by lahwaacz** [comment]
@jtpio I think that I figured it out - please review! :pray:

**Comment by jtpio** [comment]
Thanks @lahwaacz!

Looks like it needs a lint pass according to the CI logs. Running `jlpm lint` should help fix it.

**Comment by jtpio** [comment]
Looks like we also need to update the jupyterlab version here for the docs to build:

https://github.com/jupyterlab-contrib/jupyterlab-pytutor/blob/02ff7cd1372e51221b21e34226e98fdfa70fce26/docs/environment.yml#L13

**Comment by lahwaacz** [comment]
@jtpio Thanks for your comments, I managed to fix the CI workflow.

**Comment by jtpio** [comment]
Looks good on Binder :+1: 

image

File: src/index.ts [hunk 1]
```diff
@@ -109,7 +109,7 @@ export class PyTutorExtension
             src += '###############################\n';
           }
           console.log('code cell', cell_index);
-          src += `${cell.model.value.text}\n`;
+          src += `${cell.model.toJSON().source}\n`;
```
Comment (jtpio) [inline src/index.ts]: Would the following also work?

```suggestion
          src += `${cell.model.sharedModel.source}\n`;
```

Comment (lahwaacz) [reply src/index.ts]: It does work, thanks.

**Review by jtpio — APPROVED** [review]
Thanks!

**Pull request merged**